File: packages/app/src/services/software/definitions/nsis.ts
import { ok, err } from 'neverthrow';
import { XMLParser } from 'fast-xml-parser';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'nsis';
	public name = 'NSIS';
	public category = [SoftwareCategory.Development];
	public downloadName = 'nsis-setup.exe';
	public requiresAdmin = true;
	public icon = 'nsis.png';
	public homepage = 'https://nsis.sourceforge.io/Download';

	public async resolveDownloadUrl() {
		const downloadPageUrl = await this.getDownloadPageUrl();
		if (!downloadPageUrl) {
			return err(DownloadUrlResolveError.Generic);
		}

		const res = await fetch(downloadPageUrl);
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const html  = await res.text();
		const match = html.match(/<a href="(.*)" rel="nofollow">/);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(match[1]);
	}

	private async getDownloadPageUrl() {
		const res = await fetch('https://sourceforge.net/projects/nsis/rss');
		if (!res.ok) {
			return null;
		}

		const parser       = new XMLParser();
		const xml          = await res.text();
		const data         = parser.parse(xml);
		const downloadPath = data.rss.channel.item.find(
			(i: { title: string; }) => i.title.endsWith('-setup.exe')
		).title;

		return `https://sourceforge.net/settings/mirror_choices?projectname=nsis&filename=${downloadPath}&dialog=true`;
	}
}

File: src/lib/software/nsis.py
from re import compile
from PySide6.QtCore import Slot, QUrl
from src.lib.software import BaseSoftware, SoftwareCategory
from PySide6.QtNetwork import QNetworkReply, QNetworkRequest

class NSIS(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'nsis'
        self.name = 'NSIS'
        self.category = [SoftwareCategory.Development]
        self.download_name = 'nsis-setup.exe'
        self.should_cache_url = True
        self.requires_admin = True
        self.icon = 'nsis.png'
        self.homepage = 'https://nsis.sourceforge.io'

        self._major_version = 3  # Update this if version 4 is ever a thing
        self._initial_url = QUrl('https://nsis.sourceforge.io/Download')

    @Slot(QNetworkReply)
    def on_manager_finished(self, reply: QNetworkReply):
        reply.deleteLater()
        error = reply.error()
        if error != QNetworkReply.NetworkError.NoError:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            return

        html = reply.readAll().data().decode()

        if reply.url() == self._initial_url:
            version_pattern = compile(r'https://prdownloads\.sourceforge\.net/nsis/nsis-(\d+\.\d+)-setup\.exe\?download')
            match = version_pattern.search(html)
            if not match:
                self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            else:
                version = match.group(1)
                req = QNetworkRequest(f'https://sourceforge.net/settings/mirror_choices?projectname=nsis&filename=NSIS%20{self._major_version}/{version}/nsis-{version}-setup.exe&dialog=true')
                self.manager.get(req)
        else:
            download_url_pattern = compile(r'<a href="(.*)" rel="nofollow">')
            match = download_url_pattern.search(html)
            if not match:
                self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            else:
                self.url_resolved.emit(match.group(1))

    def resolve_download_url(self):
        self.manager.get(QNetworkRequest(self._initial_url))
